### DISCUSSION TITLE
Correct way to update a record with API rules

### DISCUSSION THREAD

**Discussion by osaxon** [pocketbase/pocketbase#5076] [Q&A]
I am trying to implement a feature where a user can upload an image to use as their avatar but am facing issues with the API policy.
What I'm unsure about is how to send the authorization token in the request? I am logged in as the user which I'm trying to update yet I get the 404 error.

```js   
 <form
      onSubmit={async (e) => {
        e.preventDefault();
        await db.client
          .collection("users")
          .update(
            "gf7tdqsirgu9006",
            { avatar: formData.get("file") },
            // do I set headers here?
            // or include credentials?
          );
      }}
    >
```
Whenever I try to make this request I get 404 record not found, which I believe is the default response if the request doesn't satisfy the api policy on the record.

**Comment by ganigeorgiev** [comment]
I assume that `db.client` is a JS SDK instance.

In that case the `Authorization:TOKEN` header is automatically send with the request if `db.client.authStore.token` is valid.

Double check that:
- you are sending the update request to the correct user id
- add `console.log(db.client.authStore.token)` before sending the request to verify that you are actually "logged in"
- Start the executable with `--dev` and/or find the failed request in the the _Admin UI > Logs_ for a more detailed error message.

If you want to send manually the `Authorization` header then you can submit it as 3rd optional argument with your update request (_see [Custom request options](https://github.com/pocketbase/js-sdk#custom-request-options)_), but as mentioned earlier the SDK should do this automatically for you if the auth store is populated and valid.

**Reply by osaxon** [reply]
Thanks for the quick reply. 
I think I've got it working now. Would you mind just checking my understanding on a couple of things?

Is it correct that I need to call `loadFromCookie` first in order for `authStore.isValid` to return `true`? When checking `authStore.isValid` from the form submit handler it's returning `false`.

```js
    <form
      onSubmit={(e) => {
        e.preventDefault();
        
        console.log(db.client.authStore.isValid) // false here
      }}
    >
```

The `db.client` is a class that wraps the PocketBase JS Client, with methods for logging in etc. Not sure if this is the best approach but I got this idea from [this blog post](https://blog.mahad.dev/setting-up-nextjs-and-pocketbase-for-authentication). Here is a method on the class to check if a user is authenticated. I'm using this a nextjs middleware file to redirect unauthenticated users.
The method calls the same `authStore.isValid` check but it first calls `loadFromCookie`
```js
  async isAuthenticated(cookieStore: ReadonlyRequestCookies) {
    const cookie = cookieStore.get("pb_auth");
    if (!cookie) {
      return false;
    }

    this.client.authStore.loadFromCookie(cookie?.value || "");
    console.log(this.client.authStore.isValid, "<<<< is valid"); // returned true here
    return this.client.authStore.isValid || false;
  }
```

As I don't have access to cookies within the client side React component in Next JS, I'm unable to call the `isAuthenticated` above.

I'm using tRPC in this app so the way I've managed to get around this is by passing the request cookies to the tRPC context and then I've created a mutation which has access to both the cookies and custom class I made that wraps the PB client. 

```js
const createContext = async (req: NextRequest) => {
  return createTRPCContext({
    headers: req.headers,
    pb: db,
    cookies: cookies(),
  });
};
```

```js
  test: publicProcedure.mutation(async ({ ctx }) => {
    const cookies = ctx.cookies;
    console.log("cookies", cookies);
    const isAuth = await ctx.pb.isAuthenticated(cookies);
    console.log("isAuth", isAuth); // returns true
    const updatedUser = await ctx.pb.client
      .collection("users")
      .update("wzhpt0xhsbagzps", { name: "Mabel" }); // update now works
    return updatedUser;
  }),
```

The above works, but I guess I just wanted to share these finds and sanity check if this approach seems valid.

**Reply by ganigeorgiev** [reply]
I don't think I can help you further, but if you are new to web development I don't recommend using JS SSR (especially Next.js) with PocketBase as most likely you'll messup the security. You can find more ramblings on this in https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7620032, https://github.com/pocketbase/pocketbase/discussions/4962, https://github.com/pocketbase/pocketbase/discussions/4412#discussioncomment-8575834 and probably others.
